File: src/patterns/Behavioral_Patterns/memento/ts_ver/editor/Editor.ts
import { Memento, TextArea } from "./TextStuff";

export class Editor {
    private readonly mementos: Memento[];
    private textArea: TextArea;
    
    constructor(textArea: TextArea) {
        this.mementos = [];
        this.textArea = textArea;
    }

    write(text: string) {
        this.textArea.setText(text);
        this.mementos.push(this.textArea.createMemento());
    }

    get canUndo() {
        return this.mementos.length > 0;
    }

    undo() {
        if(this.canUndo) {
            this.textArea.restore(this.mementos.pop()!);
        }
    }
}

File: src/patterns/Behavioral_Patterns/memento/java_ver/pak2/Editor.java
package pak2;

import java.util.Deque;

public class Editor {
    private Deque<TextArea.Memento> history;
    private TextArea textarea;

    public Editor(TextArea textArea) {
        this.textarea = textArea;
    }

    public void write(String text) {
        textarea.setText(text);
        history.push(textarea.takeSnapshot()); // 스냅샷을 찍어서 저장
    }

    public void undo() {
        if(!this.history.isEmpty())
            textarea.restore(this.history.pop());
    }
}
